### PR TITLE
qemu-vm: remove bootDisk, refactor using make-disk-image with better EFI support

### PR DESCRIPTION
**Pull request by RaitoBezarius**
###### Description of changes

This PR does multiple things :

####### `make-disk-image` refactoring in QEMU test infrastructure

As per https://github.com/NixOS/nixpkgs/issues/23052 — QEMU test infrastructure do not use ad-hoc code to build the boot disk image but reuse `make-disk-image` which was improved to support use-cases.

Doing this consolidates `make-disk-image` as one place to do these changes, it seems like the boot disk production code was doing old stuff with GRUB 0.97, I'm uncertain whether that should be cleaned up.

####### UEFI firmware infrastructure

As part of https://github.com/NixOS/nixpkgs/pull/187887 ; this uses consolidated UEFI firmware and variables and expose their variables to end users.

It also adds SecureBoot and System Management Mode enforcement support in the test infrastructure and the `make-disk-image` function.

EFIVARS can be modified as part of the disk image build process, enabling people to build a disk image with specific authenticated UEFI variables, e.g. SecureBoot keys.

####### Documentation

This PR adds documentation on what are the breaking changes in regard to the disk layout (`storeImage` is used only if needed at all.)

TODO/notes:

- [x] Cleanup UEFI firmware infrastructure: done in 
- [x] Cleanup the `/dev/vdb2` stuff in QEMU code.
- [x] Some tests such as installer depends on `bootDevice` being vdb or second drive.
- [x] GRUB version 1 may have some subtleties.
- [x] Assert `touchEFIVars` only if partition table type is hybrid or efi.
- [ ] Use a better error message for assert.
- [ ] Refactor in `make-disk-image` the EFI firmware infrastructure using #187887.
- [x] Simple tests:
   - [x] Legacy boot without any EFI stuff should still work.
   - [x] EFI/Hybrid boot without touching EFI vars should still work.
   - [x] EFI/Hybrid boot with touching EFI vars should still work *AND* we should be able to write/read in EFIVARS.
   - [x] Over {systemd-boot, GRUB1, GRUB2} cartesian product.
- [ ] Documentation:
  - [ ] How to test this kind of changes? What to look for?
  - [ ] Where are the breaking changes? Everything that relies on `/dev/vdb2` is now incorrect.
  - [ ] Compatibility layer? Deprecation notice?
     - [ ] Prevent users to shoot their foot: increase the number of assertions & warnings accordingly in the module.

###### For a future PR

- [ ] `makeEc2Test` is useful to initialize using an image, it would be good to extract it as a lower level primitive to start test off an already existing disk image, maybe put it directly into `makeTest` machinery or `makeTestFromImage`

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
@edolstra I think you wrote the initial `qemu-vm` module and the boot disk generation code.
It included a provision to support GRUB v0.97: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/qemu-vm.nix#L274-L275 — am I right this is not needed anymore and can be dropped in this PR?

**Comment by RaitoBezarius** [comment]
cc @roberth @lheckemann as this may affect the installer (which I tested but I think it's a touchy PR).

**Comment by RaitoBezarius** [comment]
@GrahamcOfBorg test installer.simple
@GrahamcOfBorg test simple
@GrahamcOfBorg test non-default-filesystems
@GrahamcOfBorg test swap-partition
@GrahamcOfBorg test systemd-boot.basic
@GrahamcOfBorg test systemd-boot.update
@GrahamcOfBorg test systemd-boot.switch-test

**Comment by RaitoBezarius** [comment]
@NixOS/ofborg test installer.simple
@NixOS/ofborg test simple
@NixOS/ofborg test non-default-filesystems
@NixOS/ofborg test swap-partition
@NixOS/ofborg test systemd-boot.basic
@NixOS/ofborg test systemd-boot.update
@NixOS/ofborg test systemd-boot.switch-test

**Comment by roberth** [comment]
We can run the tests on hydra before merging.

**Comment by RaitoBezarius** [comment]
@ofborg test installer.simple
@ofborg test simple
@ofborg test non-default-filesystems
@ofborg test swap-partition
@ofborg test systemd-boot.basic
@ofborg test systemd-boot.update
@ofborg test systemd-boot.switch-test

**Comment by RaitoBezarius** [comment]
> We can run the tests on hydra before merging.

That would be great, should I ping you to ask you for this?

**Comment by roberth** [comment]
I've picked the best recent commit from trunk-combined to run as a baseline. I've pushed this commit as `python-test-refactoring-baseline` for reference, as well as `python-test-refactoring` for evaluation (which is in progress right now). Could you base this branch on that commit as well?

If that leads to conflicts, we can reconsider the choice of baseline commit.

**Comment by RaitoBezarius** [comment]
> I've picked the best recent commit from trunk-combined to run as a baseline. I've pushed this commit as `python-test-refactoring-baseline` for reference, as well as `python-test-refactoring` for evaluation (which is in progress right now). Could you base this branch on that commit as well?
> 
> If that leads to conflicts, we can reconsider the choice of baseline commit.

I only need to rebase on `python-test-refactoring` branch or? I did not understand exactly what was needed for Hydra to pick up this. Should I also change the target of this PR?

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 1]
```diff
@@ -755,18 +735,17 @@ in
           '';
       };
 
-    virtualisation.efiVars =
```
Comment (roberth) [inline nixos/modules/virtualisation/qemu-vm.nix:758]: Please add a `mkRemovedOptionModule` for this option.
Alternatively, if this option could be important to users, you could add options that replicate the defaulting logic in the `make-disk-image.nix` parameters and use `mkRenamedOptionModule`.

**Comment by baloo** [comment]
https://github.com/ofborg test secureboot.basic

**Comment by roberth** [comment]
> I only need to rebase on `python-test-refactoring` branch

:+1: 

> I did not understand exactly what was needed for Hydra to pick up this.

Push to `python-test-refactoring`, but let's do it after we have a baseline eval.

> Should I also change the target of this PR?

This isn't needed. We're going to merge it into master when ok. We only use the branch for ad hoc evals on hydra. We don't have a merge train like staging, because these aren't mass-rebuilds.

**Comment by roberth** [comment]
Looks like `systemd-boot.update` fails, apart from whatever is wrong on aarch64

**Comment by RaitoBezarius** [comment]
Yeah, it fails on master too, I think these are systemd breaking changes.

Le ven. 9 sept. 2022 à 19:01, Robert Hensing ***@***.***> a
écrit :

**Comment by baloo** [comment]
> Looks like `systemd-boot.update` fails, apart from whatever is wrong on aarch64

That one is on me. It's the SMM commit that wrecks havoc in the `make-disk-image` builds. I was at least missing the overall enabling of smm in qemu (`-machine qc35,smm=on,accel=tcg`) but looks like there is something else missing. I guess I'll need to get a debug build of ovmf and get the debug console wired in.

File: nixos/lib/make-disk-image.nix [hunk 2]
```diff
@@ -368,11 +391,27 @@ let format' = format; in let
     diskImage=$out/${filename}
   '';
 
+  createEFIVars = ''
+    efiVars=$out/efi-vars.fd
+    cp ${efiVariables} $efiVars
```
Comment (rnhmjoj) [inline nixos/lib/make-disk-image.nix:396]: I've noticed a number of potential quoting issues around this module (here, line 406, 429, 430, 452, ...)
Can you run the resulting scripts through shellscript and fix them?

Comment (RaitoBezarius) [reply nixos/lib/make-disk-image.nix:396]: Of course. :)

**Comment by RaitoBezarius** [comment]
> > I only need to rebase on `python-test-refactoring` branch
> 
> +1

Done, but now, I do have… conflicts :C ?

> 
> > I did not understand exactly what was needed for Hydra to pick up this.
> 
> Push to `python-test-refactoring`, but let's do it after we have a baseline eval.

Alright!

> 
> > Should I also change the target of this PR?
> 
> This isn't needed. We're going to merge it into master when ok. We only use the branch for ad hoc evals on hydra. We don't have a merge train like staging, because these aren't mass-rebuilds.

Alright.

**Comment by RaitoBezarius** [comment]
Ah… It mass pinged everyone. :')

**Pull request closed without merging**